### PR TITLE
OCPBUGS-29453: UPSTREAM: 231: make garbage collection a runnable

### PR DESCRIPTION
**Pull request by everettraven**

### CODE REVIEW

**Review by rashmigottipati — COMMENTED** [review]
/lgtm

File: internal/garbagecollection/garbage_collector.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	if err != nil {
+		gc.Logger.Error(err, "running garbage collection")
+	}
+	if len(removed) > 0 {
+		gc.Logger.Info("removed stale cache entries", "removed entries", removed)
+	}
+
+	// Loop until context is canceled, running garbage collection
+	// at the configured interval
+	for {
```
Comment (stevekuznetsov) [inline internal/garbagecollection/garbage_collector.go:50]: FYI take a peek at the `wait` package from k8s for prior art and some of the edge cases

File: internal/garbagecollection/garbage_collector.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+var _ manager.Runnable = (*GarbageCollector)(nil)
+
+// GarbageCollector is an implementation of the manager.Runnable
+// interface for running garbage collection on the Catalog content
+// cache that is served by the catalogd HTTP server. It runs in a loop
+// and will ensure that no cache entries exist for Catalog resources
+// that no longer exist. This should only clean up cache entries that
+// were missed by the handling of a DELETE event on a Catalog resource.
```
[the remark this answers is not in the document]
Comment (everettraven) [reply internal/garbagecollection/garbage_collector.go:26]: There is a finalizer. I suspect there are no missed delete events when the controller is running. It's been long enough that I don't 100% recall why we needed garbage collection on startup, but the intention behind this comment is that this looping shouldn't be a problem and if it cleans anything up post-startup it should _only_ end up cleaning up a resource if the controller somehow missed a delete event

File: internal/garbagecollection/garbage_collector.go (hunk 1 above)
Comment (everettraven) [reply internal/garbagecollection/garbage_collector.go:50]: Can do in a future follow up. Thanks!

File: internal/garbagecollection/garbage_collector.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+				gc.Logger.Error(err, "running garbage collection")
+			}
+			if len(removed) > 0 {
+				gc.Logger.Info("removed stale cache entries", "removed entries", removed)
+			}
+		}
+	}
+}
+
+func runGarbageCollection(ctx context.Context, cachePath string, metaClient metadata.Interface) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (everettraven) [reply internal/garbagecollection/garbage_collector.go:66]: I don't think we care. This is only supposed to be garbage collection of the cached directories not the inverse.

**Pull request merged**